math/fast_power.cpp

```diff
@@ -0,0 +1,55 @@
+#include <iostream>
+#include <cstdint>
+
+/*
+    Program that computes a^b in O(logN) time.
+    It is based on formula that:
+        case1) if b is even: a^b = a^(b/2) * a^(b/2) = (a^(b/2))ˆ2
+        case2) if b is odd: a^b = a^((b-1)/2) * a^((b-1)/2) * a = (a^((b-1)/2))^2 * a
+    We can compute a^b recursively using above algorithm.
+*/
+
+int64_t fast_power_recursive(int64_t a, int64_t b) {
+    if (b == 0) return 1;
+    int64_t bottom = fast_power_recursive(a, b >> 1);
+    // Since it is integer division b/2 = (b-1)/2 where b is odd.
+    // Therefore, case2 is easily solved by integer division.
+
+    int64_t result;
+    if ((b & 1) == 0)  // case1
+        result = bottom * bottom;
+    else  // case2
+        result = bottom * bottom * a;
+    return result;
+}
+
+/*
+    Same algorithm with little different formula.
+    It still calculates in O(logN)
+*/
+int64_t fast_power_linear(int64_t a, int64_t b) {
+    int64_t result = 1;
+    while (b) {
+        if (b & 1) result = result * a;
+        a = a * a;
+        b = b >> 1;
+    }
+    return result;
+}
+
+int main() {
+    std::ios_base::sync_with_stdio(false);
+
+    std::cout << "3^5 is = " << fast_power_recursive(3, 5) << std::endl;
+    std::cout << "6^4 is = " << fast_power_recursive(6, 4) << std::endl;
+    std::cout << "2^7 is = " << fast_power_recursive(2, 7) << std::endl;
+    std::cout << "10^3 is = " << fast_power_recursive(10, 3) << std::endl;
```

Comment:
Would it be possible to create some tests of the form:
```cpp
assert (fast_power_recursive(10, 3) == 1000);
```
that would raise an error if __fast_power_recursive()__ was modified incorrectly?

Should we test zero and negative values in our tests?

Reply:
It would be possible to create tests using **assert()**.
This algorithm works for negative base but not for negative power. 
We can use the fact that aˆ(-b) = 1/(aˆb) but I do not think it is often used.

Reply:
Perhaps a Python example would help...
```

>>> 10 ** 3
1000
>>> -10 ** 3
-1000
>>> 10 ** -3
0.001
>>> -10 ** -3
-0.001
```
When run with each of these values, does the proposed algorithm return similar results?  If not, how does it fail?  Does if produce an invalid result or does it raise a helpful error?

Reply:
Perhaps tests of the form:
```cpp
assert (fast_power_recursive(10, 3) ==  std::pow(10, 3))
```

Reply:
Updated test form.